source/common/upstream/load_balancer_impl.cc

```diff
@@ -12,14 +12,91 @@ static const std::string RuntimeZoneEnabled = "upstream.zone_routing.enabled";
+
+  early_exit_zone_routing_ = earlyExitNonZoneRouting();
+  // Do not perform any calculations if we cannot perform zone routing based on non runtime params.
+  if (early_exit_zone_routing_) {
+    return;
+  }
+
+  size_t num_zones = host_set_.healthyHostsPerZone().size();
+
+  uint64_t local_percentage[num_zones];
+  calculateZonePercentage(local_host_set_->healthyHostsPerZone(), local_percentage);
+
+  uint64_t upstream_percentage[num_zones];
+  calculateZonePercentage(host_set_.healthyHostsPerZone(), upstream_percentage);
+
+  // If we have lower percent of hosts in the local cluster in the same zone,
+  // we can push all of the requests directly to upstream cluster in the same zone.
+  if ((route_directly_ = upstream_percentage[0] >= local_percentage[0])) {
+    return;
+  }
+
+  // If we cannot route all requests to the same zone, calculate what percentage can be routed.
+  // For example, if local percentage is 20% and upstream is 10%
+  // we can route only 50% of requests directly.
+  local_percent_to_route_ = upstream_percentage[0] * 10000 / local_percentage[0];
+
+  // Local zone does not have additional capacity (we have already routed what we could).
+  // Now we need to figure out how much traffic we can route cross zone and to which exact zone
+  // we should route. Percentage of requests routed cross zone to a specific zone needed be
+  // proportional to the residual capacity upstream zone has.
+  //
+  // residual_capacity contains capacity left in a given zone, we keep accumulating residual
+  // capacity to make search for sampled value easier.
+  // For example, if we have the following upstream and local percentage:
+  // local_percentage: 40000 40000 20000
+  // upstream_percentage: 25000 50000 25000
+  // Residual capacity would look like: 0 10000 5000. Now we need to sample proportionally to
+  // bucket sizes (residual capacity). For simplicity of finding where specific
+  // sampled value is, we accumulate values in residual capacity. This is what it will look like:
+  // residual_capacity: 0 10000 15000
+  // Now to find a zone to route (bucket) we could simply iterate over residual_capacity searching
+  // where sampled value is placed.
+  residual_capacity_.resize(num_zones);
+
+  // Local zone (index 0) does not have residual capacity as we have routed all we could.
+  residual_capacity_[0] = 0;
+  for (size_t i = 1; i < num_zones; ++i) {
+    // Only route to the zones that have additional capacity.
+    if (upstream_percentage[i] > local_percentage[i]) {
+      residual_capacity_[i] =
+          residual_capacity_[i - 1] + upstream_percentage[i] - local_percentage[i];
+    } else {
+      // Zone with index "i" does not have residual capacity, but we keep accumulating previous
+      // values to make search easier on the next step.
+      residual_capacity_[i] = residual_capacity_[i - 1];
+    }
+  }
+};
+
+bool LoadBalancerBase::earlyExitNonZoneRoutingRuntime() {
```

Comment:
i'll move runtime checks into tryChooseLocalZoneHosts(), we need access runtime values on each iteration if we are not in NoZoneRouting